### PR TITLE
Update RSCT operator status

### PR DESCRIPTION
**Pull request by manju956**
- Update RSCT operator status based on observed status of pods spun by the operator
- Address requirements of the issue https://github.com/ocp-power-automation/rsct-operator/issues/44
- Add required Cluster Service Version fields

### CODE REVIEW

File: bundle/manifests/rsct-operator.clusterserviceversion.yaml [hunk 1]
```diff
@@ -21,10 +21,10 @@ metadata:
         }
       ]
     capabilities: Basic Install
-    createdAt: "2024-07-10T12:35:54Z"
+    createdAt: "2024-09-27T10:44:41Z"
     operators.operatorframework.io/builder: operator-sdk-v1.34.1
     operators.operatorframework.io/project_layout: go.kubebuilder.io/v4
-  name: rsct-operator.v0.0.1
+  name: rsct-operator.v0.0.2
```
[the remark this answers is not in the document]
Comment (manju956) [reply bundle/manifests/rsct-operator.clusterserviceversion.yaml:27]: sure, will revert

File: config/crd/bases/rsct.ibm.com_rscts.yaml [hunk 2]
```diff
@@ -46,6 +46,12 @@ spec:
             type: object
           status:
             description: RSCTStatus defines the observed state of RSCT
+            properties:
+              current_status:
+                description: |-
+                  INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+                  Important: Run "make" to regenerate code after modifying this file
```
Comment (varad-ahirwadkar) [inline config/crd/bases/rsct.ibm.com_rscts.yaml:53]: Can you please update the current_status's description ?

Comment (varad-ahirwadkar) [inline config/crd/bases/rsct.ibm.com_rscts.yaml:53]: Can you please update the `current_status`'s description ?

File: internal/controller/status.go [hunk 3]
```diff
@@ -2,11 +2,84 @@ package controller
...
+	var podsStatus []string
+	for _, pod := range podList.Items {
+		switch {
+		case pod.Status.Phase == corev1.PodPending:
+			podsStatus = append(podsStatus, "PENDING")
+		case pod.Status.Phase == corev1.PodFailed:
+			podsStatus = append(podsStatus, "FAILED")
+		case pod.Status.Phase == corev1.PodRunning:
+			podsStatus = append(podsStatus, "RUNNING")
+			continue
```
Comment (mkumatag) [inline internal/controller/status.go:38]: do we need a continue statement here?

File: internal/controller/status.go [hunk 4]
```diff
@@ -2,11 +2,84 @@ package controller
...
+}
+
+// evalOperatorStatus determines operator status based on the pods status
+func evalOperatorStatus(podList *corev1.PodList) string {
+	var effectiveStatus string
+	var podsStatus []string
+	for _, pod := range podList.Items {
+		switch {
+		case pod.Status.Phase == corev1.PodPending:
+			podsStatus = append(podsStatus, "PENDING")
```
Comment (mkumatag) [inline internal/controller/status.go:33]: podstatus can be a custom data type and all the states can be a constant here

File: internal/controller/status.go [hunk 5]
```diff
@@ -2,11 +2,84 @@ package controller
...
+	// Operator status:
+	// 1. PENDING
+	// 2. RUNNING
+	// 3. PARTIALLY-RUNNING
+	// 4. FAILED
+
+	pods := &corev1.PodList{}
+
+	labelSelector := labels.SelectorFromSet(map[string]string{"app": currentDaemonSet.Name})
+	listOpts := &client.ListOptions{Namespace: rsct.Namespace, LabelSelector: labelSelector}
```
Comment (mkumatag) [inline internal/controller/status.go:69]: instead of reading the pods, wondering if we can read the state of the daemonset itself and its healthiness

File: api/v1alpha1/rsct_types.go [hunk 6]
```diff
@@ -38,6 +38,7 @@ type RSCTSpec struct {
 type RSCTStatus struct {
 	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
+	CurrentStatus *string `json:"current_status,omitempty"`
```
Comment (mkumatag) [inline api/v1alpha1/rsct_types.go:41]: state will always be current here, can we rename to state here?

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: api/v1alpha1/rsct_types.go (hunk 6 above)
Comment (manju956) [reply api/v1alpha1/rsct_types.go:41]: sure, will make the change

File: config/crd/bases/rsct.ibm.com_rscts.yaml (hunk 2 above)
Comment (manju956) [reply config/crd/bases/rsct.ibm.com_rscts.yaml:53]: sure

File: internal/controller/status.go (hunk 5 above)
Comment (manju956) [reply internal/controller/status.go:69]: the reason for choosing to inspect pod(s) status instead of daemonset state is because it carries multiple detailed states which are not useful. below is the RSCT daemonset state when its in running condition.

`status:
    currentNumberScheduled: 1
    desiredNumberScheduled: 1
    numberAvailable: 1
    numberMisscheduled: 0
    numberReady: 1
    observedGeneration: 1
    updatedNumberScheduled: 1 `

File: internal/controller/status.go (hunk 3 above)
Comment (manju956) [reply internal/controller/status.go:38]: will remove the redundant `continue` here

File: config/manager/kustomization.yaml [hunk 7]
```diff
@@ -13,4 +13,4 @@ kind: Kustomization
 images:
 - name: controller
   newName: ghcr.io/ocp-power-automation/rsct-operator
-  newTag: latest
+  newTag: 0.0.1
```
Comment (mkumatag) [inline config/manager/kustomization.yaml:16]: we can keep it to latest for now

File: go.mod [hunk 8]
```diff
@@ -16,6 +16,7 @@ require (
 
 require (
 	github.com/beorn7/perks v1.0.1 // indirect
+	github.com/cenkalti/backoff v2.2.1+incompatible // indirect
```
Comment (mkumatag) [inline go.mod:19]: why is this added as indirect all of a sudden? can you check where is this used?

File: internal/controller/status.go [hunk 9]
```diff
@@ -2,11 +2,93 @@ package controller
...
+	"k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+type PodStatus string
+
+const (
+	PENDING           PodStatus = "PENDING"
+	RUNNING           PodStatus = "RUNNING"
+	FAILED            PodStatus = "RUNNING"
```
Comment (mkumatag) [inline internal/controller/status.go:20]: why is this `"RUNNING"`? is this supposed to be `FAILED`?

File: internal/controller/status.go [hunk 10]
```diff
@@ -2,11 +2,93 @@ package controller
...
+		}
+		return false
+	}
+	return true
+}
+
+// evalOperatorStatus determines operator status based on the pods status
+func evalOperatorStatus(podList *corev1.PodList) string {
+	var effectiveStatus string
+	var podsStatus []string
```
Comment (mkumatag) [inline internal/controller/status.go:39]: can this be `[]PodStatus` and avoid string casting below

File: internal/controller/status.go [hunk 11]
```diff
@@ -2,11 +2,93 @@ package controller
...
+	}
+	return effectiveStatus
+}
+
+// updateRSCTStatus updates RSCT operator status
 func (r *RSCTReconciler) updateRSCTStatus(ctx context.Context, rsct *rsctv1alpha1.RSCT, currentDaemonSet *appsv1.DaemonSet) error {
+	// Operator status:
+	// 1. PENDING
+	// 2. RUNNING
+	// 3. PARTIALLY-RUNNING
```
Comment (mkumatag) [inline internal/controller/status.go:72]:
```suggestion
	// 3. PARTIALLY_RUNNING
```

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: config/manager/kustomization.yaml (hunk 7 above)
Comment (manju956) [reply config/manager/kustomization.yaml:16]: sure, will revert back to latest

File: internal/controller/status.go (hunk 9 above)
Comment (manju956) [reply internal/controller/status.go:20]: think its a typo, will correct it.

File: go.mod (hunk 8 above)
Comment (manju956) [reply go.mod:19]: think it was added while doing experimenting code changes and missed out from cleanup, will remove it.

File: internal/controller/status.go (hunk 10 above)
Comment (manju956) [reply internal/controller/status.go:39]: makes sense. will do

**Comment by mkumatag** [comment]
btw: there is something wrong with your GitHub user configuration, please fix that

**Review by mkumatag — APPROVED** [review]
/lgtm

**Comment by manju956** [comment]
> btw: there is something wrong with your GitHub user configuration, please fix that
> 
> image

@mkumatag  May I know what's wrong with user configuration?

**Comment by mkumatag** [comment]
> May I know what's wrong with user configuration?

I think you have fixed it, see a proper link to your commit..

**Review by mkumatag — APPROVED** [review]
/lgtm

**Comment by manju956** [comment]
> /lgtm

@mkumatag can you remove `do-not-merge` label so that we can merge the PR?
